samples/extensions/extended_dynamic_state2/extended_dynamic_state2.h

```diff
@@ -0,0 +1,169 @@
+/* Copyright (c) 2023, Mobica Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 the "License";
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "api_vulkan_sample.h"
+#include "scene_graph/components/pbr_material.h"
+
+class ExtendedDynamicState2 : public ApiVulkanSample
+{
+  public:
+	typedef struct ModelDynamicParam
+	{
+		bool depth_bias         = false;
+		bool rasterizer_discard = false;
+	} ModelDynamicParam;
+
+	struct
+	{
+		bool                           tessellation = false;
+		float                          tess_factor  = 1.0;
+		float                          patch_control_points_float{3.0f};
+		int                            patch_control_points{3};
+		std::vector<ModelDynamicParam> objects;
+		int                            selected_obj     = 0;
+		bool                           selection_active = true;
+		bool                           time_tick        = false;
+	} gui_settings;
+
+	struct UBOBAS
+	{
+		glm::mat4 projection;
+		glm::mat4 view;
+		glm::vec4 ambientLightColor = glm::vec4(1.f, 1.f, 1.f, 0.1f);
+		glm::vec4 lightPosition     = glm::vec4(-3.0f, -8.0f, 6.0f, -1.0f);
+		glm::vec4 lightColor        = glm::vec4(1.0f, 1.0f, 1.0f, 1.0f);
+		float     lightIntensity    = 50.0f;
+	} ubo_baseline;
+
+	struct UBOTESS
+	{
+		glm::mat4 projection;
+		glm::mat4 modelview;
+		float     tessellation_factor = 1.0f;
+	} ubo_tess;
+
+	struct UBOBG
+	{
+		glm::mat4 projection;
+		glm::mat4 background_modelview;
+
+	} ubo_background;
```

Comment:
The three UBO structs have the first two members in common.
Maybe introduce a new UBO with just those two members to be used in each case, and a special UBO for the baseline and the tessellation part?

Reply:
UBO refactored, common uniform buffer introduced (used in all 3 pipelines), shaders for tessallation are now simplified, thanks for the suggestion, changes merged